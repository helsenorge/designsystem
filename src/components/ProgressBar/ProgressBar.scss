$progress-bar-border-radius: 1rem;

.progress-bar {
  background-color: #e9e9e9;
  padding: 0.1875rem;
  border-radius: $progress-bar-border-radius;
  overflow: hidden;
  border-top: 1px solid #e1e1e1;
  border-right: 1px solid #e1e1e1;
  border-left: 1px solid #e1e1e1;
  border-bottom: 1px solid #b7b7b7;
  &__indicator {
    background-color: #009657;
    border-top-left-radius: $progress-bar-border-radius;
    border-bottom-left-radius: $progress-bar-border-radius;
    transition: width 0.5s ease-in-out;
    animation: progress-bar 1s;
    @keyframes progress-bar {
      0% {
        width: 0;
      }
    }
    &--complete {
      border-top-right-radius: $progress-bar-border-radius;
      border-bottom-right-radius: $progress-bar-border-radius;
    }
    &__value {
      position: relative;
      &--outside {
        right: -100%;
        padding-left: 0.75rem;
      }
      &--inside {
        color: white;
        text-align: right;
        padding-right: 0.75rem;
      }
    }
  }
}
