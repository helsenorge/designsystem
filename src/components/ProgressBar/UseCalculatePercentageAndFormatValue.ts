import {useState, useEffect} from 'react';

interface CalculatePercentageAndFormatValueProps {
  min: number;
  max: number;
  value: number;
}

export function useCalculatePercentageAndFormatValue({
  min,
  max,
  value,
}: CalculatePercentageAndFormatValueProps): [number, number] {
  const [[percentage, formatValue], setPercentageAndFormatValue] = useState([0, 0]);
  useEffect(() => {
    if (value <= min) {
      setPercentageAndFormatValue([0, min]);
    } else if (value >= max) {
      setPercentageAndFormatValue([100, max]);
    } else {
      setPercentageAndFormatValue([((value - min) / (max - min)) * 100, value]);
    }
  }, [min, max, value]);
  return [percentage, formatValue];
}
