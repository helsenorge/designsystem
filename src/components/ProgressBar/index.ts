export {ProgressBar, useCalculatePercentageAndFormatValue} from './ProgressBar';
