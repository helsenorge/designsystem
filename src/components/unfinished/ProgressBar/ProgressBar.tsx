/*  This file has been converted to TypeScript
 When making changes to this file please make them
 in the file with a .ts or .tsx extension located in
 the Common/src/toolkit folder structure
 Then run the command "npm run build-toolkit" to generate the jsx file
 Please check both files into TFS */

import * as React from 'react';
import * as classNames from 'classnames';

export interface ProgressbarProps {
  min: number;
  max: number;
  value: number;
  formatValue?: (value: number) => string;
}

export default class Progressbar extends React.Component<ProgressbarProps, {}> {
  public static defaultProps: ProgressbarProps = {
    min: 0,
    max: 100,
    value: 0,
    formatValue: undefined,
  };

  calculateCompletedPercent(): number {
    const {value, min, max}: ProgressbarProps = this.props;
    if (value < min) {
      return 0;
    }
    if (value > max) {
      return 100;
    }
    const diff: number = max - min;
    return ((value - min) / diff) * 100;
  }

  shouldValueBeInside(): boolean {
    return this.calculateCompletedPercent() >= 50;
  }

  formatValue(): string {
    const {value, min, max, formatValue}: ProgressbarProps = this.props;
    let valueToFormat: number = value;
    if (value < min) {
      valueToFormat = min;
    }
    if (value > max) {
      valueToFormat = max;
    }
    if (formatValue) {
      return formatValue(valueToFormat);
    }
    return valueToFormat + '';
  }

  render(): JSX.Element {
    const completeClasses: string = classNames('complete', {
      full: this.calculateCompletedPercent() >= 100,
    });
    const classes: string = classNames('value', {
      inside: this.shouldValueBeInside(),
      outside: !this.shouldValueBeInside(),
    });
    return (
      <div className="atom_progressbar">
        <div className="track">
          <div className={completeClasses} style={{width: `${this.calculateCompletedPercent()}%`}}>
            <div className={classes}>{this.formatValue()}</div>
          </div>
        </div>
      </div>
    );
  }
}
