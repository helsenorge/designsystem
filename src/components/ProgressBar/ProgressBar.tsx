import React from 'react';
import cn from 'classnames';
import styles from './ProgressBar.scss';
import {useCalculatePercentageAndFormatValue} from './UseCalculatePercentageAndFormatValue';

interface ProgressBarProps extends React.HTMLProps<HTMLDivElement> {
  min?: number;
  max?: number;
  value: number;
  formatValue?: (value: number) => string;
}

const ProgressBar = React.forwardRef<HTMLDivElement, ProgressBarProps>(
  ({min = 0, max = 100, value, formatValue = value => `${value}`, ...restProps}, ref): JSX.Element => {
    const [percentage, formattedValue] = useCalculatePercentageAndFormatValue({min, max, value});

    function isValueInside(): boolean {
      return percentage >= 50;
    }

    function isComplete(): boolean {
      return percentage >= 100;
    }

    const indicatorClasses = cn(styles['progress-bar__indicator'], {
      [styles['progress-bar__indicator--complete']]: isComplete(),
    });

    const valueClasses = cn(
      styles['progress-bar__indicator__value'],
      {[styles['progress-bar__indicator__value--inside']]: isValueInside()},
      {[styles['progress-bar__indicator__value--outside']]: !isValueInside()},
    );

    return (
      <div ref={ref} className={styles['progress-bar']}>
        <div className={indicatorClasses} style={{width: `${percentage}%`}}>
          <div className={valueClasses}>{formatValue(formattedValue)}</div>
        </div>
      </div>
    );
  },
);

export {ProgressBar};
