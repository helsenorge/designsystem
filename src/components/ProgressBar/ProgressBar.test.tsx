import React from 'react';
import {render} from 'react-testing-library';
import {ProgressBar} from './ProgressBar';

test('renders correctly', (): void => {
  const {container} = render(<ProgressBar value={0}>ProgressBar</ProgressBar>);
  expect(container).toMatchSnapshot();
});
