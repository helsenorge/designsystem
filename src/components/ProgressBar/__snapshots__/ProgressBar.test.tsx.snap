// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="progress-bar"
  >
    <div
      class="progress-bar__indicator"
      style="width: 0%;"
    >
      <div
        class="progress-bar__indicator__value progress-bar__indicator__value--outside"
      >
        0
      </div>
    </div>
  </div>
</div>
`;
